src/System.Memory/src/System/Buffers/SequenceReader.cs

```diff
@@ -120,6 +120,58 @@ public SequenceReader(ReadOnlySequence<T> sequence)
             }
         }
 
+        /// <summary>
+        /// Peeks at the next value at specific offset without advancing the reader.
+        /// </summary>
+        /// <param name="offset">The offset from current position.</param>
+        /// <param name="value">The next value or default if at the end.</param>
+        /// <returns>False if at the end of the reader.</returns>
+        public readonly bool TryPeek(long offset, out T value)
+        {
+            if (offset < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException_OffsetOutOfRange();
+
+            // If we've got data and offset is not out of bounds
+            if (!_moreData || Remaining <= offset)
+            {
+                value = default;
+                return false;
+            }
+
+            // If offset doesn't fall inside current segment move to next until we find correct one
+            if ((CurrentSpanIndex + offset) <= CurrentSpan.Length - 1)
+            {
+                value = CurrentSpan[CurrentSpanIndex + (int)offset];
```

Comment:
Add the debug assert when casting a long to an int where it is known to be safe (i.e. the cast won't lose any info).

```C#
Debug.Assert(offset <= int.MaxValue);
```